addons/token-exchange/blue_secret_controller.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	queueKeyFn := func(obj runtime.Object) string {
+		key, err := cache.MetaNamespaceKeyFunc(obj)
+		if err != nil {
+			return ""
+		}
+		return key
+	}
+
+	eventFilterFn := func(obj interface{}) bool {
+		if s, ok := obj.(*corev1.Secret); ok {
+			if s.Type == RookType && strings.Contains(s.ObjectMeta.Name, blueSecretMatchString) {
+				return true
+			}
+		}
+		return false
+	}
+
+	return factory.New().
+		WithFilteredEventsInformersQueueKeyFunc(queueKeyFn, eventFilterFn, spokeSecretInformers.Informer()).
+		WithSync(c.sync).
+		ToController(fmt.Sprintf("blue-secret-%s-controller", TokenExchangeName), recorder)
+}
+
+// sync gets invoked when the informer
+func (c *blueSecretTokenExchangeAgentController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	key := syncCtx.QueueKey()
+	klog.Infof("Blue secret controller: Reconciling addon deploy %q", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore addon whose key is not in format: namespace/name
+		return nil
+	}
+
+	secret, err := getSecret(c.spokeSecretLister, name, namespace)
+	if err != nil {
+		return fmt.Errorf("failed to get the secret %q in namespace %q in managed cluster. Error %v", name, namespace, err)
+	}
+
+	sc, err := c.getStorageClusterFromRookSecret(secret)
+	if err != nil {
+		return fmt.Errorf("failed to get the storage cluster name from the sercet %q in namespace %q in managed cluster. Error %v", name, namespace, err)
+	}
+
+	newSecret, err := createBlueSecret(secret, sc, c.clusterName)
+	if err != nil {
+		return fmt.Errorf("Failed to create secret from the managed cluster secret %q from namespace %v for the hub cluster in namespace %q err: %v", secret.Name, secret.Namespace, c.clusterName, err)
+	}
+
+	err = createSecret(c.hubKubeClient, c.recorder, &newSecret)
+	if err != nil {
+		return fmt.Errorf("Failed to sync managed cluster secret %q from namespace %v to the hub cluster in namespace %q err: %v", secret.Name, secret.Namespace, c.clusterName, err)
+	}
+
+	klog.Infof("successfully synced managed cluster secret %q from namespace %v to the hub cluster in namespace %q", secret.Name, secret.Namespace, c.clusterName)
+
+	return nil
+}
+
+func (c *blueSecretTokenExchangeAgentController) getStorageClusterFromRookSecret(secret *corev1.Secret) (storageCluster string, err error) {
+	for _, v := range secret.ObjectMeta.OwnerReferences {
+		if v.Kind != "CephCluster" {
+			continue
+		}
+
+		rclient, err := rookclient.NewForConfig(c.kubeconfig)
+		if err != nil {
+			return "", fmt.Errorf("Unable to create a rook client err: %v", err)
+		}
+
+		found, err := rclient.CephV1().CephClusters(secret.Namespace).Get(context.TODO(), v.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", fmt.Errorf("Unable to fetch ceph cluster err: %v", err)
+		}
+		klog.Infof("cephclusteris %+v", found)
+
+		for _, owner := range found.ObjectMeta.OwnerReferences {
+			klog.Infof("cephcluster object meta is %+v", owner)
```

Comment:
was this log for testing? Don't think we need to log this.

Reply:
Done, thanks!